src/net/sourceforge/kolmafia/textui/langserver/textdocumentservice/AshTextDocumentService.java

```diff
@@ -30,14 +37,55 @@ public final void setCapabilities(final ServerCapabilities capabilities) {
   }
 
   @Override
-  public void didOpen(DidOpenTextDocumentParams params) {}
+  public void didOpen(DidOpenTextDocumentParams params) {
+    this.parent.executor.execute(
+        () -> {
+          TextDocumentItem document = params.getTextDocument();
+
+          File file = FilesMonitor.UriToFile(document.getUri());
+
+          this.parent.monitor.updateFile(file, document.getText(), document.getVersion());
+        });
+  }
 
   @Override
-  public void didChange(DidChangeTextDocumentParams params) {}
+  public void didChange(DidChangeTextDocumentParams params) {
+    this.parent.executor.execute(
+        () -> {
+          VersionedTextDocumentIdentifier document = params.getTextDocument();
+          List<TextDocumentContentChangeEvent> changes = params.getContentChanges();
+
+          if (changes.size() == 0) {
+            // Nothing to see here
+            return;
+          }
+
+          File file = FilesMonitor.UriToFile(document.getUri());
+
+          // We don't support incremental changes, so we expect the client
+          // to put the whole file's content in a single TextDocumentContentChangeEvent
+          this.parent.monitor.updateFile(file, changes.get(0).getText(), document.getVersion());
```

Comment:
Could you raise an error if multiple changes are raised, as it's not supported? As you only handle the first, and assume it's the whole text.

Reply:
Since this is a notification, the client won't get a response. As such, raising an error is pointless.

Besides, more than a single change simply means "undefined behaviour". It's not stated that it needs to result in an error.
We said that we supported `TextDocumentSyncKind.Full`. This means we just look at the first element.
They didn't put any? IndexOutOfBoundsException and we silently move on.
They put more? We don't care.